File: flutter-wrapper/lib/proxy/proxy_node.dart
class ProxyNode {
  final String name;
  final String type;
  final String server;
  final int port;
  final String cipher;
  final String password;
  final bool udp;

  const ProxyNode({
    required this.name,
    required this.type,
    required this.server,
    required this.port,
    required this.cipher,
    required this.password,
    this.udp = false,
  });

  factory ProxyNode.fromJson(Map<String, dynamic> json) {
    final type = json['type'] as String;
    if (type != 'ss') {
      throw FormatException('unsupported proxy type: $type');
    }
    final port = json['port'] as int;
    if (port < 1 || port > 65535) {
      throw FormatException('port out of range [1, 65535]: $port');
    }
    return ProxyNode(
      name: json['name'] as String,
      type: type,
      server: json['server'] as String,
      port: port,
      cipher: json['cipher'] as String,
      password: json['password'] as String,
      udp: (json['udp'] as bool?) ?? false,
    );
  }

  Map<String, dynamic> toJson() => {
        'name': name,
        'type': type,
        'server': server,
        'port': port,
        'cipher': cipher,
        'password': password,
        'udp': udp,
      };

  @override
  String toString() => 'ProxyNode($name $server:$port)';
}

File: electron-wrapper/proxy/proxy-node.js
'use strict';

/**
 * Defensive validator for ProxyNode-shaped objects fetched from OSS at
 * runtime. The backend already normalizes builtinProxies before injection,
 * but OSS responses are foreign data — must be validated before use.
 *
 * Returns a normalized ProxyNode object on success; throws TypeError on
 * any shape problem.
 */
function validateNode(o) {
  if (!o || typeof o !== 'object') {
    throw new TypeError('node must be object');
  }
  for (const key of ['name', 'type', 'server', 'port', 'cipher', 'password']) {
    if (!(key in o)) {
      throw new TypeError(`node missing field ${key}`);
    }
  }
  if (o.type !== 'ss') {
    throw new TypeError(`unsupported type ${o.type}`);
  }
  const port = typeof o.port === 'string' ? parseInt(o.port, 10) : o.port;
  if (!Number.isInteger(port) || port < 1 || port > 65535) {
    throw new TypeError(`port out of range: ${o.port}`);
  }
  return {
    name: String(o.name || `${o.server}:${port}`),
    type: 'ss',
    server: String(o.server),
    port,
    cipher: String(o.cipher),
    password: String(o.password),
    udp: Boolean(o.udp),
  };
}

module.exports = { validateNode };
